Draft: ghc: Make GHC more robust on AArch64/Darwin

bgamari opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Currently the `llc` and `opt` tools are not specified explicitly, meaning that
`ghc` using whichever executables it finds on `PATH`. Here we specify absolute
paths to these tools in the `configure` invocation, ensuring that these needn't be in PATH for the compiler to
find them.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

bgamari committed:
ghc: Configure LLC and OPT explicitly

Set LLC and OPT explicitly to the absolute paths to the respective LLVM
utilities, ensuring that these needn't be in PATH for the compiler to
find them.

bgamari committed:
ghc: Pass --target to cc on Darwin

This ensures that we compile for the correct target, even when Rosetta
is available.

See GHC #20162.

sternenseemann commented:
Can we do something like this for all required tools? I'd be delighted if we could get rid of `runtimeDeps` for all platforms tbh.

sternenseemann commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -231,6 +234,9 @@ stdenv.mkDerivation (rec {
     "--with-iconv-includes=${libiconv}/include"
     "--with-iconv-libraries=${libiconv}/lib"
+  ] ++ lib.optionals stdenv.isDarwin [
+    "CFLAGS=--target=${stdenv.hostPlatform.config}"
```
I think the nixpkgs cc wrapper always passes `-target` and it should have no effect passing it a second time actually:

<https://github.com/NixOS/nixpkgs/blob/2eb8695646429ed33c060c868bf4be5f464257d2/pkgs/build-support/cc-wrapper/default.nix#L301>

AFAIK we don't bypass the cc-wrapper for GHC anywhere.

maralorn commented:
We must not forget to mirror those changes into the ghc 9.x derivations.

bgamari committed:
ghc: Ensure that clang is hermetic

GHC uses clang for its assembler on Darwin. However, previously we
failed to specify `CLANG`, meaning that the default value of
`CLANG=clang` would be used.

sternenseemann commented:
@bgamari I almost forgot about this PR, but ended up doing roughly the same changes in <https://github.com/NixOS/nixpkgs/pull/147260/commits/17b8b5a4dc5a49d83d508cb66c6ac2f0bb9fa39b>. This indeed solves the LLVM-related confusion when compiling GHC and also makes it more robust in general (e. g. the GHC LLVM no longer leaks into the `llvm-hs` build).

Additionally I've taken the opportunity to pass `OTOOL` and `INSTALL_NAME_TOOL` which solves the issues we had at some point on darwin more cleanly.

Currently still waiting on the rebuild, but looking good so far. Will be merged as part of #147260, but I'm not quite sure if we can backport it to 21.11.

sternenseemann commented:
Probably.
